pyee/base.py

```diff
@@ -0,0 +1,362 @@
+# -*- coding: utf-8 -*-
+
+from collections import defaultdict, OrderedDict
+from threading import Lock
+from typing import Any, Dict, Generic, List, Optional, Protocol, Tuple, TypeVar, Union
+
+from typing_extensions import Literal
+
+
+class PyeeException(Exception):
+    """An exception internal to pyee."""
+
+
+class PyeeError(PyeeException):
+    """An error internal to pyee."""
+
+
+Event = TypeVar(name="Event", contravariant=True)
+ErrorEvent = Literal["error"]
+NewListenerEvent = Literal["new_listener"]
+InternalEvent = Union[ErrorEvent, NewListenerEvent]
+
+Arg = TypeVar(name="Arg", contravariant=True)
+Kwarg = TypeVar(name="Kwarg", contravariant=True)
+
+
+class AnyHandlerP(Protocol):
+    """
+    A handler with any args or kwargs. This type is used
+    to capture the type of a handler emitted by the
+    ``new_listener`` event.
+    """
+
+    def __call__(self, *args: Any, **kwargs: Any) -> Any:
+        ...
+
+
+class HandlerP(Protocol[Event, Arg, Kwarg]):
+    """
+    An event handler.
+    """
+
+    def __call__(
+        self,
+        *args: Union[Arg, Exception, Event, InternalEvent, "AnyHandlerP"],
+        **kwargs: Kwarg,
+    ) -> Any:
+        ...
+
+
+IEvent = TypeVar(name="IEvent")
+IArg = TypeVar(name="IArg")
+IKwarg = TypeVar(name="IKwarg")
+
+
+class DecoratorP(Protocol[IEvent, IArg, IKwarg]):
+    """
+    A handler decorator function.
+    """
+
+    def __call__(
+        self, f: HandlerP[IEvent, IArg, IKwarg]
+    ) -> HandlerP[IEvent, IArg, IKwarg]:
+        ...
+
+
+class EventEmitter(Generic[Event, Arg, Kwarg]):
+    """The base event emitter class. All other event emitters inherit from
+    this class.
+
+    Most events are registered with an emitter via the ``on`` and ``once``
+    methods, and fired with the ``emit`` method. However, pyee event emitters
+    have two *special* events:
+
+    - ``new_listener``: Fires whenever a new listener is created. Listeners for
+      this event do not fire upon their own creation.
+
+    - ``error``: When emitted raises an Exception by default, behavior can be
+      overridden by attaching callback to the event.
+
+      For example::
+
+          @ee.on('error')
+          def on_error(message):
+              logging.err(message)
+
+          ee.emit('error', Exception('something blew up'))
+
+    All callbacks are handled in a synchronous, blocking manner. As in node.js,
+    raised exceptions are not automatically handled for you---you must catch
+    your own exceptions, and treat them accordingly.
+    """
+
+    def __init__(self) -> None:
+        self._events: Dict[
+            Union[Event, InternalEvent],
+            "OrderedDict[HandlerP[Event, Arg, Kwarg], AnyHandlerP]",
+        ] = defaultdict(OrderedDict)
+        self._lock: Lock = Lock()
+
+    def on(
```

Comment:
I think what I want to do is make type-safe variants of these method:

```python
def on_user(
    event: UserEvent
    f: HandlerP[Arg, Kwarg]
) -> HandlerP[Arg, Kwarg]: ...

def on_error(
    event: ErrorEvent,
    f: HandlerrP[Exception, None]
) -> HandlerP[Exception, None]: ...

def on_new_listener(
    event: NewListenerEvent,
    f:HandleP[Union[UserEvent, ErrorEvent, NewListenerEvent, HandlerP[Any, Any]]]
) -> HandlerP{[Union[UserEvent, ErrorEvent, NewListenerEvent, HandlerP[Any, Any]]]: ...
```

Then `on` becomes:

```python
def add_listener(
    self,
    event: Union[Event, ErrorEvent, NewListenerEvent],
    f: HandlerP[Any, Any]
) -> Handler[Any, Any]: ...

That latter method would call the other safer methods. It might have to pull some dirty tricks, though I think the `Any` types here will do the hard work.

Reply:
I couldn't make this work! Oh well!